llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -1230,7 +1234,92 @@ static bool isUnsupportedAMDGPUAddrspace(Value *Addr) {
   return false;
 }
 
+static bool isUnsupportedSPIRAddrspace(Value *Addr, Function *Func) {
+  Type *PtrTy = cast<PointerType>(Addr->getType()->getScalarType());
+  if (PtrTy->getPointerAddressSpace() != 0) {
+    // Skip SPIR-V built-in varibles
+    StringRef Name;
+    if (auto *CE = dyn_cast<ConstantExpr>(Addr)) {
+      const auto OpCode = CE->getOpcode();
+      if (OpCode == Instruction::GetElementPtr ||
+          OpCode == Instruction::AddrSpaceCast) {
+        Name = CE->getOperand(0)->getName();
+      }
+    } else {
+      Name = Addr->getName();
+    }
+    if (Name.starts_with("__spirv_BuiltIn")) {
+      return true;
+    }
+  } else {
+    // FIXME: Check kernel argments instead of value's name
+    auto Name = Addr->getName();
+    if (Name == "_arg_NumWorkItems" || Name == "_arg_KernelFunc") {
+      return true;
+    }
+  }
+  return false;
+}
+
+static GlobalVariable *GetOrCreateGlobalString(Module &M, StringRef Name,
+                                               StringRef String,
+                                               unsigned AddressSpace) {
+  static std::unordered_map<std::string, GlobalVariable *> GlobalStringMap;
```

Comment:
It's better to change this static variable to member variable of AddressSanitizer pass.

Reply:
why?

Reply:
According to the code logic, I think GlobalStringMap should belong to AddressSanitizer pass, it would be easier to maintain (the lifetime of static variable is hard to control). BTW why not use llvm::DenseMap?

> If do so, will modify a lot of code

Maybe you can do this change later.

Reply:
Done. Move it into AddressSanitizer class.